### DISCUSSION TITLE
Adjoint gradients for dCLdalpha far from 2pi

### DISCUSSION THREAD

**Discussion by albisb** [mdolab/dafoam#503] [General]
Hi all,

I am trying to run a flat plate at an angle of attack for validation purposes. First of all, I am trying to implement as a design variable the angle of attack. For that, I need to make it different to adjusting the inlet velocity direction, since my final case will involve pure rotation of only a part of the total geometry and hence it is not possible for me to use the inlet conditions for this. 

I have implemented the angle of attack using a twist function inside the runScript_v2, as given by the OpenMDAO/MACH Aero:
def twist(val, geo):
    for i in range(nRefAxPts):
        geo.rot_z["bodyAxis"].coef[i] = val[i-1]*180/np.pi
(And I convert it to degrees because I am working with radiants)

To test it, I am trying to optimize the CL. This way, I know that the lift slope should be somehow close to 2pi. After implementing it, while the value of the variables in optLog.txt is consistent with the angle of attack I am getting and the optimization result is pure rotation, when I am checking the adjoint gradients in totalDerivHist.txt, these have a value of around 2, far from the 2pi I want to obtain, while a lift slope CL(AoA)/AoA is close to 6, and I am working at small angles (2°) to ensure that the lift polar is still linear.

Does anyone have a suggestion why the adjoint gradient could be different to the objective value? Or any suggestion of a better way to implement an AoA variable without changing the inlet conditions?

Thanks

**Comment by uttamcadambi07** [comment]
Hi! 


Before I post, I want to understand something - why do you expect the total derivative to be close to 2pi? I mean, the total derivative is the ratio of the change in the CL to the change in the angle (assuming the design variable is the angle of attack and the objective function is the lift co-efficient). Either way, the gradient will only progressively take you to 2pi, not in one go. I am not sure if my understanding is right, so feel free to correct me .


I have implemented a translation operation to perform something similar (but in my case, I was trying to infer the location of the origin using some reference data).

I do not know if it will be useful but I am posting the translate function here. See if you can fix your twist function based on how I have coded the translate function 

```
# =============================================================================
# Design variable setup
# =============================================================================
def translate(val, geo):
    # Get the current control points
    C = geo.extractCoef("c4")
    print(C)
    # Calculate the translation distances in x and y directions
    translate_x = val[0]
    translate_y = val[1]
    
    # Create a translation vector for the x and y directions
    translation_vector = np.array([translate_x, translate_y, 0.0])
    
    # Apply the translation to each control point
    for i in range(len(C)):
        C[i, :] += translation_vector
        print(C[i,:])
    
    # Restore the modified control points to the geometry
    geo.restoreCoef(C, "c4")        

DVGeo = DVGeometry("./FFD/wingFFD.xyz")
nrefaxpts = DVGeo.addRefAxis("c4", xFraction=0.25, alignIndex="k")
DVGeo.addGlobalDV("translation", func=translate, value=[-0.09, -0.1], lower=[-0.2, -0.2], upper=[-0.1, 0.2], scale=1)
daOptions["designVar"]["translation"] = {"designVarType": "FFD"}

```


I was getting the design variable to move realistic distances, but my gradients were REALLY small. To give you an example, here is a design variable output from one of the iterations 

```
"Optimization_Iteration_004":
{
    "translation": [ -1.683798455472551e-01, -8.831047851266045e-02 ]
},
```
The initial design variable values are given as 

```
"Optimization_Iteration_001":
{
    "translation": [ -1.009999902000000e-01, -1.000000000000000e-01 ]
},
```

And here is the corresponding gradient at that iteration

```
"Optimization_Iteration_004":
{
    "FI": 
    {
        "translation": [ 6.328271240363392e-15, -1.076916333886402e-14 ]
    }
```

**Reply by albisb** [reply]
Hi,

Thanks a lot for the detailed description of your translation function. Regarding the value of the gradient, I would say it should be near the theoretical 2pi. Assuming the plate is within the linear region (which it is, because it is only at 2°), we can expect a gradient deltaCl/deltaAoA to be close to 2pi, unless the delta is too large (this for finite differences). To ensure that this is not the case for the adjoint, I tried changing the daOptions::adjPartDerivFDStep::State/FFD to assess sensitivity (changing one order of magnitude up and down) but didn't see any difference in the gradients. 
This are the sensitivities I am getting:
"Optimization_Iteration_001":
{
    "CL":
    {
        "alpha": [ 3.145480860531862e-02 ],
        "twist": [ -2.061153446686908e+00, -2.119879790677042e+00 ]
    }
},
It is also weird there is one for alpha since I removed it, although it crashed without so I defined a blank alpha:
def alpha(val, geo): return 0

Finally, the objective function (CL) values I am getting are:
Objective Functions:
{'CL': 0.1573193019231554, 'fail': False}

Which for a geometrical angle of attack (the one at the starting point) this would result in a lift slope of 4.5 (I am using a very coarse mesh which is very diffusive for testing it, when refining the mesh lift slope tends to 2pi). However, the sensitivity to the "twist" angle is around 2.

Regarding the implementation of the rotation function, I can see the flat plate is rotating correctly, as expected. If I only had to add some concerns about it, it seems that the mesh deformation also affects a bit the inlet, as seen in the image below after one step of optimization. However, I tested different domain sizes and the results don't seem to be very sensitive to it, so I'm not particularly worried about it:
![image](https://github.com/mdolab/dafoam/assets/145017905/8b8f1a99-e636-47df-8e3e-19f8a465f100)
![image](https://github.com/mdolab/dafoam/assets/145017905/a63cdc96-0cfc-4306-b6fe-212c5d26bc89)

**Comment by friedenhe** [comment]
Regarding your first post, the way you define the twist variable is not quite right. You should have only one twist variable, so please use sth like this:

```
def twist(val, geo):
    for i in range(nRefAxPts):
        geo.rot_z["bodyAxis"].coef[i] = val[0]*180/np.pi
```

Then, you don't run an optimization to verify your point. Instead, run the flow and adjoint once and check dCL/dTwist.

BTW: there is no need to use aoa for this case because your incoming flow is always in the x direction, and the lift is always in the y direction.

**Reply by albisb** [reply]
Ok, I re-implemented the twist function, so now the "two tips" are controlled by the same variable. Now, I saw that with a finer mesh, the gradient tends to converge to the "finite differences" calculation.
Coarse mesh
- CL = 0.157 @ 2° --> deltaCLdeltaTwist = 4.5
- dCldTwist = 4.2
Fine mesh:
- CL = 0.197 @ 2° --> deltaCLdeltaTwist = 5.65
- dCldTwist = 5.66
So it seems to work as expected. 
Other than that, I need to include alpha, otherwise it crashes and complains about "alpha" variable not defined, as if it was something standard that the code expects. So I had to define and include alpha (although in this case, the function I defined should keep it zero).
For instance, if I remove all the references to alpha of the runScript_v2 (definition of function alpha, specifying daOptions["designVar"]["alpha"] = ... , adding it as a global DV and removing entry "alphaName" = "alpha") it will crash saying:
--> FOAM FATAL IO ERROR: 
Entry 'alphaName' not found in dictionary ".part1"

file: .part1

    From function bool Foam::dictionary::readEntry(const Foam::word&, T&, Foam::keyType::option, bool) const [with T = Foam::word]
    in file /home/dafoamuser/dafoam/OpenFOAM/OpenFOAM-v1812/src/OpenFOAM/lnInclude/dictionaryTemplates.C at line 201.

FOAM exiting

Or if I add back alphaName entry:
--> FOAM FATAL IO ERROR: 
Entry 'alpha' not found in dictionary ".designVar"

file: .designVar

This is not a problem since I can define a dummy alpha, but just to inform about it.

Many thanks

**Comment by friedenhe** [comment]
This is because you set CD and CL's directions as `parallel/normalToFlow`. This setting requires you setup an alpha variable to determine the flow direction. You can set the CD and CL's directions as `fixedDirection` so there is no need to set alpha.

**Comment by albisb** [comment]
You're right, I checked again a simulation using fixedDirection and removing the alpha variable and it didn't crash. Also the results and gradients are exactly the same, so it works perfectly.
Many thanks
